### PR TITLE
Qsignature - add 2 profiler classes to report on qsignature positions and qsignature output

### PR DESCRIPTION
**Pull request by holmeso**
# Description

In an attempt to better understand the positions within the `qsignature` positions file, and also the outputs generated by `qsignature`, 2 profiling classes have been added.
The `VcfProfiler` looks at a vcf file that has been generated by `qsignature` and provide various stats that are hoped will allow the user to glean the type of input BAM file that the vcf file was generated from.
It outputs a json file that is a serialised version of the SigVcfMeta class.
The `jackson-databind` project has been added to the `qsignatue` `build.gradle` file as a dependency. Version 2.6.7.1 was chosen as it is also used by another jar within `adamajava`.

The `PositionsProfiler` class looks at the `qsignature` positions file (vcf) and outputs some statistics of interest into the log file. It is not anticipated that this class will be often used.

## Type of change

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

# How Has This Been Tested?

Existing unit tests pass, new unit tests have been added.

# Are WDL Updates Required?

As this is new code, there are no wdl updates that are required.
It is envisaged that the `VcfProfiler` will be called by workflow level wdl scripts in the future.

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by ChristinaXu2017** [comment]
The current main usage only have two modes, this two new modes should be added into as well
USAGE = usage: qsignature [org.qcmg.sig.SignatureGeneratorBespoke | org.qcmg.sig.CompareGenotypeBespoke] [-options]

**Comment by ChristinaXu2017** [comment]
Whether unit test for PositionsProfiler exists?  Are there any real data tests for these two new classes?

**Comment by holmeso** [comment]
> Whether unit test for PositionsProfiler exists? Are there any real data tests for these two new classes?

There is no unit test for the PositionsProfiler class - it doesn't produce any output other than a log file.
Once the germline GRCh38 process uses the VcfProfiler class, a regression test will be setup to examine the generated output

**Comment by holmeso** [comment]
> The current main usage only have two modes, this two new modes should be added into as well
> USAGE = usage: qsignature [org.qcmg.sig.SignatureGeneratorBespoke | org.qcmg.sig.CompareGenotypeBespoke] [-options]
> 
> Where is the qsignature command? I can't find any batch script.

Nice spot.
I've updated the usage statements accordingly.

**Comment by ChristinaXu2017** [comment]
> There is no unit test for the PositionsProfiler class - it doesn't produce any output other than a log file.
> Once the germline GRCh38 process uses the VcfProfiler class, a regression test will be setup to examine the generated output

OK, I see the PositionsProfiler class is mainly for internal use.  However we still should make sure it provides the right stats from input to the log file. Unit tests still required, otherwise mark it somewhere, such as it is a draft and accept mistakes.

**Comment by holmeso** [comment]
> > There is no unit test for the PositionsProfiler class - it doesn't produce any output other than a log file.
> > Once the germline GRCh38 process uses the VcfProfiler class, a regression test will be setup to examine the generated output
> 
> OK, I see the PositionsProfiler class is mainly for internal use. However we still should make sure it provides the right stats from input to the log file. Unit tests still required, otherwise mark it somewhere, such as it is a draft and accept mistakes.

I've updated the `PositionsProfiler` class documentation telling users not to trust the log output along with a link to your comment.

File: qsignature/src/org/qcmg/sig/messages.properties [hunk 1]
```diff
@@ -1,6 +1,8 @@
-USAGE = usage: qsignature [org.qcmg.sig.SignatureGeneratorBespoke | org.qcmg.sig.CompareGenotypeBespoke] [-options]
-COMPARE_USAGE = usage: qsignature org.qcmg.sig.Compare -log <log_file> -dir <directory in which to search for .qsig.xml files> -output <output xml file>
-GENERATOR_USAGE = usage: qsignature org.qcmg.sig.SignatureGeneratorBespoke -snpPositions <file containing positions of interest> -illuminaArraysDesign <Illumina arrays design document - contains list of snp ids and whether they should be complemented> -input <bam or snp chip file, or folder containing bams or snp chip files> -output <output directory where qsig.xml files will be written to> -log <log_file>
+USAGE = usage: qsignature.jar [org.qcmg.sig.SignatureGeneratorBespoke | org.qcmg.sig.Compare | org.qcmg.sig.VcfProfiler] [-options]
```
Comment (ChristinaXu2017) [inline qsignature/src/org/qcmg/sig/messages.properties:1]: Should the usage start with  "USAGE = usage: java -cp qsignature.jar ..." ?

File: qsignature/src/org/qcmg/sig/messages.properties [hunk 2]
```diff
@@ -1,6 +1,8 @@
...
+COMPARE_USAGE = usage: qsignature.jar org.qcmg.sig.Compare -log <log_file> -dir <directory in which to search for .qsig.xml files> -output <output xml file>
+GENERATOR_USAGE = usage: qsignature.jar org.qcmg.sig.SignatureGeneratorBespoke -snpPositions <file containing positions of interest> -illuminaArraysDesign <Illumina arrays design document - contains list of snp ids and whether they should be complemented> -input <bam or snp chip file, or folder containing bams or snp chip files> -output <output directory where qsig.xml files will be written to> -log <log_file>
```
Comment (ChristinaXu2017) [inline qsignature/src/org/qcmg/sig/messages.properties:3]: same here

File: qsignature/src/org/qcmg/sig/messages.properties [hunk 3]
```diff
@@ -1,6 +1,8 @@
...
+POSITIONS_PROFILER_USAGE = usage: qsignature.jar org.qcmg.sig.PositionsProfiler -log <log_file> -snpPositions <snp_positions_vcf_file>
+VCF_PROFILER_USAGE = usage: qsignature.jar org.qcmg.sig.VcfProfiler -log <log_file> -input <qsignature_generated_vcf_file> -snpPositions <snp_positions_vcf_file> -output <output_file.json>
```
Comment (ChristinaXu2017) [inline qsignature/src/org/qcmg/sig/messages.properties:5]: same here

**Review by ChristinaXu2017 — COMMENTED** [review]
> I've updated the `PositionsProfiler` class documentation telling users not to trust the log output along with a link to your comment.

looks good!

File: qsignature/src/org/qcmg/sig/messages.properties (hunk 1 above)
Comment (holmeso) [reply qsignature/src/org/qcmg/sig/messages.properties:1]: Looks like some duplication of `usage` here.
Does JP's document have any guidance here?

Comment (holmeso) [reply qsignature/src/org/qcmg/sig/messages.properties:1]: All our tools have this (well, all the ones I have checked). If we want to change this, it should be to all tools and in a seperate PR.

Comment (ChristinaXu2017) [reply qsignature/src/org/qcmg/sig/messages.properties:1]: In JP's modified document, all usage includes "java -jar <tool>.jar ..." or "java -cp <tool>.jar ...".  We used to have a batch script to run the tool, but not now.  However we never have "usage: <tool>.jar ..."
It is better to make it consistent with our document. There are many tools already followed JP's suggestion, please refer to q3tiledaligner, qprofiler, q3vcftools, qbamfilter, qmito.

Comment (holmeso) [reply qsignature/src/org/qcmg/sig/messages.properties]: > In JP's modified document, all usage uses "java -jar tool.jar ..." or "java -cp tool.jar ...". We used to have a batch script to run the tool, but not now. However we never have "usage: tool.jar ..."
> It is better to make it consistent with our document. There are many tools already followed JP's suggestion, please refer to q3tiledaligner, qprofiler, q3vcftools, qbamfilter, qmito.

Ah - apologies - I thought you were referring to "usage" in the usage.
Updated

**Review by ChristinaXu2017 — APPROVED** [review]
GOOD!

**Pull request merged**